papers/incorrectness24/incorrectness24.tex

```diff
@@ -0,0 +1,366 @@
+      \infer{}{\Gamma \vdash \MATH.\ABS : (\neg\NUMBER \fun \ERROR)} \\[\bigskipamount]
+      \infer{}{\Gamma \vdash \MATH.\ABS : \neg{\FUNCTION} \dashv \emptyset}
+      \end{array}
+      \infer{
+        \Gamma \vdash \STRING.\LOWER : (\neg\STRING \fun \ERROR) \\
+        \Gamma \vdash \STRING.\LOWER : \neg{\FUNCTION} \dashv \emptyset \\
+        \Gamma \vdash x : \neg{\STRING} \dashv (x : \neg\STRING) \\
+        \mbox{(warn since $\Gamma \vdash \STRING.\LOWER : \UNKNOWN \fun (\neg\NUMBER \cup \ERROR)$)}
+      }{\Gamma \vdash \STRING.\LOWER(x) : \neg{\NUMBER} \dashv (x : \neg\STRING)}
+    }{\Gamma \vdash (\MATH.\ABS(\STRING.\LOWER(x)) : \NEVER \dashv (x : \neg\STRING)}
+  \]
+  \caption{Example warning}
+  \label{fig:example}
+\end{figure*}
+
+In Figure~\ref{fig:ctxtgen} we provide some of the inference rules for
+context synthesis, and the warnings that it
+produces.  These are run after type inference, so they can assume that
+all code is fully typed.
+
+In the judgment $\Gamma \vdash M : T \dashv
+\Delta$, the type context $\Gamma$ is the usual \emph{checked} type
+context and $\Delta$ is the \emph{synthesized} context used to predict
+run-time errors (following the terminology of bidirectional
+typing~\cite{BidirectionalTyping}).
+
+\begin{conjecture}\label{conj:complete}%
+If $\Gamma \vdash M : T \dashv \Delta, x:U$ and $\sigma$ is a closing
+substitution where $\sigma(x) : U$ and $M[\sigma] \rightarrow^* v$,
+then $v : T$.
+\end{conjecture}
+
+\begin{corollary}\label{cor:complete}%
+If $\Gamma \vdash M : \NEVER \dashv \Delta, x:\UNKNOWN$ and $\sigma$ is a closing
+substitution, then $M[\sigma]$ does not terminate successfully.
+\end{corollary}
+
+\section{Checked functions}
+
+The crucial aspect of this type system is that we have a type $\ERROR$
+inhabited by no values, and by expressions which may throw a run-time exception.
+(This is essentially a very simple type and effect system~\cite{Nielson1999}
+with one effect.)
+
+The rule for function application $M(N)$
+has two dependencies on the type for $M$:
+\[\begin{array}{c}
+  \Gamma \vdash M : (S \fun \ERROR)
+  \\[\jot]
+  \Gamma \vdash M : (\UNKNOWN \fun (T \cup \ERROR))
+\end{array}\]
+Since Luau is based on semantic subtyping~\cite{GF05:GentleIntroduction,Jef22:SemanticSubtyping} and supports
+intersection types, this is equivalent to asking for $M$ to be an
+overloaded function, where one overload has argument type $\UNKNOWN$, and
+one has result type $\ERROR$. For example:
+\[
+  \MATH.\ABS : (\NUMBER \fun \NUMBER) \cap (\neg\NUMBER \fun \ERROR)
+\]
+and so (by subsumption):
+\[\begin{array}{c}
+  \MATH.\ABS : (\neg\NUMBER \fun \ERROR)
+  \\[\jot]
+  \MATH.\ABS : (\UNKNOWN \fun (\NUMBER \cup \ERROR))
+\end{array}\]
+This is a common enough idiom it is worth naming it:
+we call any function of type 
+\[
+  (S \fun T) \cap (\neg S \fun \ERROR)
+\]
+a \emph{checked} function, since it performs a run-time check
+on its argument. They are called \emph{strong functions}
```

Comment:
https://elixir-lang.org/blog/2023/09/20/strong-arrows-gradual-typing/#:~:text=If%20a%20function%20is%20strong,not%20need%20further%20runtime%20checks! 
In the blogpost these are called Strong Arrows

Reply:
Good catch. Fixed.